동기화 메모장 [Step2-2] 앨리, 제인, 조이

Allie-e opened the pull request:
안녕하세요 수박 🍉
@soo941226
앨리, 제인, 조이 입니다.

말씀드린대로 Step2의 전반적인 설계를 수정해서 다시 PR요청 드립니다!
delegate pattern활용 및 역할 분리 등 최대한 data를 한 곳에서 관리하도록 해보았습니다.
부족한 부분이 많겠지만 이번에도 리뷰 잘 부탁드립니다! 😄

감사합니다!

# STEP 2 : 코어데이터 DB 구현

## 2-1 고민했던 것

### 설계

이번 설계의 목적은 아래와 같습니다.

- 데이터를 한곳에서 관리하기
- ListViewController / DetailViewController 에서 model 관련 로직을 최대한 덜어내기
- MVC 디자인 패턴으로 구조 짜기

처음에는 SplitViewController를 이용해서 한 곳에서 data를 관리해주는 방법으로 고민을 했습니다. 하지만, SplitViewController도 ListViewController / DetailViewController와 마찬가지로 ViewController라는 점에서 그 또한 데이터를 관리해주는 역할을 맡기에 부적절하는 생각이 들었습니다. 

고민끝에, DataManager의 인스턴스를 SplitViewController가 가지고 있고, ListVC / DetailVC 에게 전달해주는 방식으로 구현하기로 결정을 했습니다. 

기존에 ListVC / DetailVC에서 Model 관련 로직이 많았던 부분은 모두 DataManager 안으로 옮겨주었습니다. (ex `if indexPath.row < dataManager.memos.count` 이런식으로 조건을 확인하는 부분)

ListVC와 DetailVC는 각각 MemoDataManagerListDelegate, MemoDataManagerDetailDelegate를 채택합니다.

DataManager는 데이터에 변화가 일어날 때 listDelegate / detailDelegate에게 일을 시킵니다.

예를들어 

유저가 셀을 삭제 → 삭제 이벤트를 받은 ListVC가 DataManager의 메모삭제 메서드를 호출 → DataManager는 조건을 확인하여 listDelegate / detailDelegate에게 적절한 일을 시킴 

결과적으로 ListVC / DetailVC은 SplitViewController으로부터 받은 DataManager를 가지고 있고, DataManager는 listDelegate / detailDelegate를 가지고 소통을 하는 구조입니다.

## indexPath를 활용하여 하나의 메서드로 ListVC와 DetailVC에서 메모 삭제

메모를 지울 때 indexPath를 활용해서 지워야 해서 Core Data에서 선택된 메모의 indexPath를 가져오는 방법에 대해 고민하였습니다.

- ListVC의 메모 목록에서 스와이프해서 삭제할때는 선택된 indexPath의 정보를 같이 전달해줄 수 있지만,
- DetailCV의 메모 상세페이지에서 더보기 버튼으로 삭제시에는 선택된 셀의 indexPath를 알 수 없어서

indexPath가 있으면 그대로 사용하고, 없으면 listVC의 selectedCellIndex에서 선택된 셀의 인덱스를 가져와서 listVC의 셀도 지우고 detailVC의 텍스트도 지워주도록 구성하였습니다. 

```swift
final class MemoListViewController: UIViewController {
		func tableView(_ tableView: UITableView,
		               trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath
		) -> UISwipeActionsConfiguration? {
		    let deleteAction = UIContextualAction(style: .destructive, title: "Delete") { _, _, _ in
		        self.dataManager.deleteSelectedMemo(at: **indexPath**)
		    }
		    ...
		}
}
```

```swift
final class MemoDetailViewController: UIViewController {
		private func showDeleteAlert(_ sender: UIBarButtonItem) {
		    ...
		    let delete = UIAlertAction(title: "삭제", style: .destructive) { _ in
		        self.dataManager.deleteSelectedMemo()
		    }
		    ...
		}
}
```

```swift
final class MemoDataManager {
		...
		func deleteSelectedMemo(at indexPath: IndexPath? = nil) {
		    let selectedIndexPath: IndexPath?
		    if indexPath != nil {
		        selectedIndexPath = indexPath
		    } else {
		        selectedIndexPath = listDelegate?.selectedCellIndex
		    }
		    
		    guard let selectedIndexPath = selectedIndexPath else {
		        return
		    }
		    let deletedMemo = memos[selectedIndexPath.row]
		    deleteMemo(id: deletedMemo.id)
		    listDelegate?.deleteCell(at: selectedIndexPath)
		    
		    if selectedIndexPath.row < memos.count {
		        let memo = memos[selectedIndexPath.row]
		        detailDelegate?.showTextView(with: memo)
		        listDelegate?.selectNextCell(at: selectedIndexPath)
		    } else {
		        detailDelegate?.showIneditableTextView()
		    }
		}
		...
}
```

## 2-2 질문드리고 싶은 부분

## 앱 첫 실행 화면에서 첫번째 셀이 선택되지 않는 문제

<img src="https://i.imgur.com/fNxtYVr.png" width="50%" height="50%">

정상적으로 작동하는 모습

### 문제 상황

앱을 실행했을 때, selectFirstMemo()가 실행되면서 첫번째 셀이 선택되고 해당 셀의 내용을 detailView에 보여줘야합니다.

원래는 MemoListViewController의 viewDidLoad() 에서 selectFirstMemo()를 호출주었습니다.

그런데 List의 셀은 선택이 되지만, 선택된 셀의 상세페이지는 보이지 않는 문제가 발생했습니다. 

반대로 MemoDetailViewController에서 viewDidLoad() 에서 selectFirstMemo()를 호출하면 

선택된 셀의 상세페이지는 보이지만 List의 셀은 선택이 되지 않는 문제가 발생합니다. 

### 원인

문제의 원인은 selectFirstMemo()를 호출하는 VC의 delegate 등록만 작동해서 발생하는 것이었습니다.

- ListVC에서 호출하면 listDelegate만 등록되고 detailDelegate는 nil → 첫번째 셀이 선택되지만, 상세메모는 보이지 않음
- DetailVC에서 호출하면 detailDelegate만 등록되고 listDelegate는 nil → 첫번째 상세메모는 보이지만 셀 선택이 안됨

```swift
extension MemoDataManager {
    func selectFirstMemo() {
        if memos.isEmpty == false {
            listDelegate?.setupRowSelection()
            detailDelegate?.showTextView(with: memos[0])
        }
    }
}
```


|ListViewController의 viewDidLoad에서 호출하는 모습|detailDelegate가 nil인 모습|
|:---:|:---:|
|<img src="https://i.imgur.com/HOFGyRr.png" width="50%" height="50%">|<img src="https://i.imgur.com/LLxnl5I.png" width="50%" height="50%">|


해당 메서드를 호출하지 않으면 delegate임을 명시하더라도 무시되는 현상을 확인했습니다. 

좌측 사진은 ListViewController의 `viewDidLoad()`에서 `selectFirstMemo()`를 호출하고 있는 상태고, 

listDelegate / detailDelegate를 출력했을 때 `detailDelegate` 만 nil 이 나오는것을 확인할 수 있었습니다.

### 해결

listVC의 `viewDidAppear`에서 `dataManager.selectFirstMemo()` 호출하는 방식으로 해결했습니다. 

```swift
private let dataManager = MemoDataManager()

private lazy var listViewController = MemoListViewController(dataManager: dataManager)
private lazy var detailViewController = MemoDetailViewController(dataManager: dataManager)

override func viewDidAppear(_ animated: Bool) {
    super.viewDidAppear(animated)
    dataManager.selectFirstMemo()
}
```

### 궁금한 점

해결은 했지만 왜 정확히 viewDidAppear에서는 되는건지는 알아내지 못했습니다. ㅠㅠ 

dataManager.selectFirstMemo()를 viewDidLoad에서 호출하면 왜 저희가 예상한대로 작동을 하지 않는지 궁금합니다.

## 메모를 삭제할 때 viewContext를 save하기전에 memos 배열에서도 삭제하지 않으면 Crash가 나는 문제


|saveViewContext를 먼저호출|saveViewContext를 나중에 호출|
|:---:|:---:|
|<img src="https://i.imgur.com/1n9pMDt.jpg" width="50%" height="50%">|<img src="https://i.imgur.com/beG1DkD.jpg" width="50%" height="50%">|


### 문제 상황

처음에는 view context 에서 메모를 삭제하고나서 `viewContext.delete(memoToDelete)`

바로 view context를 save 한 후 `saveViewContext()`

MemoDataManager가 가지고 있는 memos 배열에서도 삭제하였습니다.

그랬더니 Crash가 나며 앱이 중지되었습니다.

### 해결

MemoDataManager가 가지고 있는 memos 배열에서도 삭제해준 후 view context를 save 하는 방법으로 

이렇게 순서를 바꿔서 해결하긴 했는데 Crash가 나는 이유가 궁금합니다.

### 궁금한 점

저희가 생각하기로는 viewContext에서는 삭제가 되었지만 MemoDataManager가 가지고 있는 memos 배열에는 데이터가 남아있어서 viewContext를 저장할때 정보가 일치하지 않아서 생기는 문제라고 이해했는데 저희 생각이 적절한지 궁금합니다!
```swift
final class MemoDataManager {
	...
  var memos = [Memo]()
	...
	func deleteMemo(id: UUID) {
      let request = Memo.fetchRequest()
      let predicate = NSPredicate(format: "id == %@", id.uuidString)
      request.predicate = predicate
      
      do {
          let memos = try viewContext.fetch(request)
          guard let memoToDelete = memos.first else {
              return
          }
          viewContext.delete(memoToDelete)
          let memoToDeleteInArray = self.memos.filter { $0.id == id }
          self.memos = self.memos.filter { memoToDeleteInArray.contains($0) == false }
          saveViewContext()
      } catch {
          print(error.localizedDescription)
      }
  }
}
```

na-young-kwon committed:
style: swiftLint 적용

siwonkim0 committed:
feat: Storyboard 삭제 후 SceneDelegate에서 SplitViewController를 root로 설정 #1

- UISplitViewController 타입의 SplitViewController 생성하여 초기 세팅

Allie-e committed:
feat: Memo, MemoDataManager 구현 #3

- Memo extension으로 attributedString을 사용하여 body와 lastModified를 하나의
문자열로 병합
- String extension으로 문자열의 길이를 제한하는 truncated() 메소드 구현
- MemoDataManager 타입을 구현하여 JSON data를 파싱하여 가지도록 함

na-young-kwon committed:
feat: MemoListViewController의 테이블뷰 구현 #1

- UITableViewDatasource 구현
- NavigationBar구현
- 메모 추가 기능 구현
- MemoDataManager의 shared프로퍼티 static으로 변경
- Memo의 subtitle 글자제한을 30 -> 40으로 변경

siwonkim0 committed:
feat: MemoDetailViewController의 textView, List->Detail 데이터 전달 구현 #2

- MemoDetailViewControllerDelegate 생성
- List뷰컨의 delegate를 detail 뷰컨으로 지정하여 detail뷰컨의 textView 업데이트
- List뷰컨에서 새로운 메모 추가시 제일 첫번째 셀이 선택되도록 구현

Allie-e committed:
feat: MemoListViewControllerDelegate 구현 #2

- detailVC의 delegate를 listVC으로 지정하여 textView에 수정이 있을 때 cell에
반영하도록 구현
- add 버튼 클릭했을 때, 첫번째 cell이 선택되도록 수정
- collection extension 추가하여 안전하게 배열 조회

na-young-kwon committed:
feat: 메모 삭제 기능 구현 #2

- tableView Delegate의 메서드인 trailingSwipeActionsConfigurationForRowAt을 사용하여 셀을 삭제하도록 구현
- didEndEditingRowAt 메서드를 통해 셀삭제 후에도 셀선택을 유지하도록 구현

siwonkim0 committed:
feat: keyboard 관련 기능 구현 #2

- NotificationCenter를 활용하여 textView에 맞게 키보드를 올리는 기능 구현
- UITapGestureRecognizer를 활용하여 빈 뷰를 선택하면 키보드를 내리는 기능 구현

Allie-e committed:
feat: CoreData 적용 #4

- JSON Data Model 및 관련 코드 삭제
- MemoDataManager의 persistentContainer 생성 및 saveViewContexte() 메소드 구현

na-young-kwon committed:
fix: 맨 마지막 셀 삭제시 index오류가 나는 버그 수정 #6

- indexPath와 메모의 개수를 비교하여 indexPath가 작거나 같을때만
 셀을 삭제하도록 수정
- deleteMemo 함수 분리
- title, body가 모두 비어있을 때 textView.text를 emptyString으로 수정

siwonkim0 committed:
feat: Detail 뷰컨의 더보기 버튼에 ActionSheet 형태의 Alert 구현 #5

- popoverPresentationController를 사용하여 아이패드 환경에서도 Alert이 띄워지도록 구현

Allie-e committed:
feat: viewMoreButtonTapped() 세부사항 구현 #5

- share 버튼 눌렀을 때, activityView 보여지도록 추가
- delete 버튼 눌렀을 때, alert 보여지도록 추가

na-young-kwon committed:
feat: 메모셀을 스와이프했을 때 공유버튼 구현 #5

- 공유버튼을 누르면 activityView가 스플릿뷰의 중심에 나타나도록 구현

siwonkim0 committed:
fix: 메모 수정 중에 코어데이터에 여러번 업데이트되는 현상 수정 #6

- 기존에는 코어데이터의 entity인 Memo 타입을 텍스트뷰가 수정될때마다 업데이트하였으나,
viewContext가 여러번 저장되어 title, body, lastModified만 업데이트하는 방식으로 수정

Allie-e committed:
feat: MemoDataManager의 deleteMemo() 메소드 추가 #4

- MemoDetailViewControllerDelegate에 showEmptyTextView() 메소드 추가하여 마지막
셀 삭제될 때 빈 textView를 보여주도록 구현

na-young-kwon committed:
feat: 더보기 버튼을 눌렀을 때 셀을 삭제하는 기능 구현 #5

- MemoListViewControllerDelegate에 deleteTableViewCell메서드 추가
- 삭제버튼을 눌렀을 때 delegate가 테이블뷰 셀을 삭제하도록 구

 Please enter the commit message for your changes. Lines starting

siwonkim0 committed:
refactor: MemoDataManager 접근제어 추가, 로직 수정 #7

Allie-e committed:
refactor: MemoListViewController 전제척인 refactoring #7

na-young-kwon committed:
refactor: MemoListViewController 전체적인 refactoring #7

- 전체적으로 final 키워드 붙임
- 외부로 노출할 필요가 없는 메서드 private 메서드로 변경
- 마지막 메모가 삭제된 후 textView의 수정이 불가능한 버그 수정

siwonkim0 committed:
refactor: MemoDetailViewController 메서드 분리 및 재배치 #7

Allie-e committed:
refactor: 호출순서에 따른 코드 위치 수정 및 refactoring #7

na-young-kwon committed:
refactor: MemoDataManager의 deleteMemo() 메서드 수정

- 인덴트 수정
- memo -> memos로 네이밍 변경
- 타입명시 삭제

siwonkim0 committed:
docs: STEP 1, 2 README 작성

Allie-e committed:
docs: Update README.md

na-young-kwon committed:
refactor: SplitViewController 리팩토링 #7

- 상속이 불필요하기 때문에 final 클래스로 변경
- setupChildView -> assignChlidViewController로 메서드명 수정

siwonkim0 committed:
refactor: Memo 타입의 subtitle 연산 프로퍼티 기능 분리 #7

Allie-e committed:
refactor: custom splitViewController 삭제  #7

- splitVC 초기설정 sceneDelegate에서 설정
- MemoDetailViewController 내부 강제 언래핑 수정

na-young-kwon committed:
refactor: hideKeyboard -> registerGestureRecognizer로 메서드명 변경 #7

siwonkim0 committed:
refactor: 데이터가 있는지 체크하여 초기 화면 세팅하는 기능 한번에 관리 #7

- listViewController에서 MemoDataManager.shared.isEmpty 가 아닌 경우 (데이터가 있는 경우) 0번째 셀을 선택하는 것과, 0번째 데이터를 memoDetail에 보여주는 기능을 한번에 관리하도록 수정

Allie-e committed:
refactor: deleteMemo() 메소드 내부 로직 변경  #7

- 메모를 지울 때, MemoDataManager의 memos와 viewContexxt에서 각각 삭제해주던
방식을 하나의 메소드 내부에서 처리하도록 수정
- 필요없는 attribute keyword 삭제
- loadPersistentContainer에서 error 발생시 에러 메세지 프린트하도록 수정

na-young-kwon committed:
refactor: memos의 초기값을 세팅하는 메서드의 호출위치 변경 #7

- memos배열의 초기값을 세팅하는 setMemosArray 메서드 추가
- appDelegate에서 didFinishLaunchingWithOptions 시점에 호출
- memos private으로 변경
- MemoListViewController에서 테이블뷰 초기세팅을 담당하는
 setupTableViewInitialSetting 메서드 생성
- MemoDetailViewController에서 textView delegate를 설정하는 로직
 메서드로 분리

Allie-e committed:
refactor: MemoDataManager의 memos 접근제어 삭제 #7

na-young-kwon committed:
refactor: SplitViewController 추가 #8

- AppDelegate/SceneDelegate에 있던 코드 SplitViewController로 이동

siwonkim0 committed:
refactor: MemoDataManager 싱글턴 제거, 델리게이트 설정 #8

- MemoDataManagerListDelegate, MemoDataManagerDetailDelegate 프로토콜 생성
- splitVC에 있던 loadPersistentContainer, setMemosArray 메서드 MemoDataManager의 이니셜라이저로 위치 변경

Allie-e committed:
refactor: MemoDataManger delegate 구현 #8

- MemoDataManger의 listDelegate, detailDelegate를 각각 ListVC와 DetailVC로 설정
- splitVC가 MemoDataManger를 갖고, 하위의 VC를 생성할 때 이니셜라이저를 통해
 주입
- 앱이 실행될 때, 첫번째 셀이 선택되도록 selectFirstMemo() 메소드 구현 및 ListVC의 viewDidAppear() 시점에 호출

na-young-kwon committed:
refactor: 메모를 더하는 기능 수정#8

- MemoDataManagerDetailDelegate에 showEmptyTextView메서드 추가
- MemoDataManager에 addNewMemo메서드 추가

siwonkim0 committed:
refactor: MemoDataManager에서 deleteSelectedMemo 구현 #8

- 기존에 MemoListVC에서 하던 셀 스와이프후 삭제 버튼이 눌리면 데이터를 삭제하는 로직을
MemoDataManager로 이동

Allie-e committed:
refactor: 더보기 버튼 클릭했을 때 셀 삭제기능 구현 #8

- 더보기 버튼의 삭제를 클릭했을 때  MemoDataManager의 deleteSelectedMemo()
 메서드 내부에서 셀을 삭제하도록 로직 수정
- 사용하지않는 deleteMemo() 메서드 삭제

na-young-kwon committed:
refactor: 메모를 업데이트하는 로직 수정 #8

- updateEditedMemo메서드 추가
- MemoDataManagerListDelegate프로토콜의 메서드 네이밍 수
- 사용하지 않는 프로토콜 삭

siwonkim0 committed:
refactor: 셀을 클릭했을때 메모 내용이 보이도록 구현 #8

- MemoDataManager showMemo() 메서드에서 선택된 셀의 인덱스를 이용해 memo를 가져와서 detailVC에 전달
- 네이밍 수정
- 필요없는 메서드 삭제

Allie-e committed:
style: 전반적인 코드 스타일 수정 #8

- 메서드 네이밍 수정
- 접근제어 수정
- 네임스페이스 생성하여 매직넘버 관리
- 코드 위치 수정

na-young-kwon committed:
refactor: 오타 수정 #8

soo941226 commented on CloudNotes/CloudNotes/Extension/UIViewController+Extension.swift:
```diff
@@ -1,10 +1,3 @@
-//
-//  Created by Siwon Kim on 2022/02/15.
-//
-
```
이러한 주석은 왜 삭제하시는지 궁금합니다. 파일을 만들면 왜 이런 주석이 생긴다고 생각하시나요?

Allie-e replied:
쓸데없는 주석들은 지워주는 것이 깔끔하다고 생각해서 삭제해주었습니다! 코멘트를 보고 생각해보니 파일 생성시 해당 주석들이 생기면서 누가, 언제 만들었는지 확인할 수 있을 것 같아요. 다만, 짝프로그래밍을 진행하면서 생성자나 생성시점이 프로젝트를 진행하기에 필요한 정보라는 생각은 들지 않습니다. 수박은 이러한 주석들을 남겨놓으시는 편인가요? 그렇다면 그 이유도 궁금합니다! 😊😊

soo941226 replied:
오 납득이 되는 이유네요!👍👍
전 혼자할 때도 남겨놓는 편인데 꼭 커밋을 보지 않더라도 어떤 날짜 순으로 파일이 생성되었는지를 알 수 있어서, 이걸 근거로 코드가 어떤 순서로 작성되었는지 알 수 있을 것 같아요. 특히 누군가에게 보여줘야 할 코드라면 이러한 내용이 꽤 유용하다고 생각을 하고 있습니다.
협업을 할 때에는 특히 담당자가 누구인지 찾을 수 있는 정보라는 생각이 드네요!

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoDetailViewController.swift:
```diff
@@ -34,8 +34,8 @@ final class MemoDetailViewController: UIViewController {
+        guard let userInfo: NSDictionary = sender.userInfo as NSDictionary?,
+              let keyboardFrame = userInfo.value(forKey: UIResponder.keyboardFrameEndUserInfoKey) as? NSValue else {
             return
         }
```
안전한 처리 좋습니다!👍

soo941226 commented on CloudNotes/CloudNotes/Model/MemoDataManager.swift:
```diff
@@ -20,6 +20,10 @@ final class MemoDataManager {
+    func setMemosArray() {
+        memos = fetchMemos()
+    }
+    
```
MemoDataManager가 [[Memo]]를 가지고 있나요?
MemosArray는 오해의 소지가 있을 거 같아요.

na-young-kwon replied:
생각도 못했는데, 복수형이 중복으로 들어가있네요,,ㅎㅎ
setMemos로 수정했습니다.

soo941226 commented on CloudNotes/CloudNotes/Model/MemoDataManager.swift:
```diff
@@ -2,7 +2,7 @@ import CoreData
 final class MemoDataManager {
     static let shared = MemoDataManager(modelName: "CloudNotes")
-    private var memos = [Memo]()
+    var memos = [Memo]()
```
이건 좀 의아한 커밋이군요🤔
외부에서 접근하는 요소가 없는데 private를 뺀 건 왜일까요?

siwonkim0 replied:
현재 listVC에서 memos에 접근해서 데이터의 개수를 알아오거나, memo를 가져오는 부분이 남아있었습니다. 이 부분도 메서드를 만들어서 memos에 직접 접근하는 부분이 없도록 수정하겠습니다!

soo941226 commented on CloudNotes/CloudNotes/Model/MemoDataManager.swift:
```diff
@@ -111,3 +114,14 @@ extension MemoDataManager {
+            detailDelegate?.showTextView(with: memos[0])
+        }
+    }
+}
```
DataManager가 UI 요소를 알고 이에 대한 일을 하게 되는 건 조금 과한 책임이 아닐까 싶어요. Manager는 DataManager니까 UI는 자기 담당이 아니라고 하지 않을까요?🤔

na-young-kwon replied:
이 부분도 리팩토링 하면서 UI관련들을 모두 덜어주었습니다!

soo941226 commented:
안녕하세요~~ 엘리, 제인, 조이!🙇‍♂️
 
 일단 고생 많으셨습니다! 이렇게 빨리 다시 PR을 주실 거라고는 진짜 생각 못했는데 정말 엄청나게 열심히 노력하신 것 같아요. 코드에서 많은 고민이 느껴졌고 앞서 이야기했던 부분이 많이 개선된 것으로 읽었습니다! 존경합니다 😄👍👍
 
 질문주신 내용에 대한 대답에 앞서, 제가 이전에 이야기한 것들 중 잘 전달이 되지 않은 부분도 있는 것 같아서 하나 다시 이야기를 드려보고자 합니다. **고치라는 말은 아니고, 잘못되었다는 말은 더더욱 아니지만,** 여러분이 정말 고민을 하셨는지 피드백이 없어서 확인차 이야기를 남깁니다!

 먼저 Debug view hierarchy를 통해 현재의 계층구조가 어떻게 표현이 되는지 눈으로 확인을 해보시고, 모델과 연결되어야 하는 컨트롤러가 누구일지 고민을 해보시면 좋겠습니다. 특히 현재에는 MemoListViewController, MemoDetailViewController가 DataManager를 전달받은 뒤, 직접 메시지를 주고 받는 구조로 보이는데요.
 제 생각에는 MemoListViewController와 MemoDetailViewController가 DataManager를 알아야할까 라는 생각이 듭니다🤔
 DataManager를 알아야하는 타입이 많음으로 인해서, DataManager에 발생하는 일들이 어떻게 관리가 되고 있는지 파악하기 힘들어지고 처리가 분산되어있다고 느껴져요. 처음에 이를 갖게 되는 SplitViewController가 이러한 일들을 모두 담당할 수 있을 것 같습니다. SplitViewController는 DataManager와 childViewControllers와 메시지를 주고 받고, 각각의 childViewController들은 splitViewController와만 메시지를 주고 받으면 좋지 않을까 하는 생각이 듭니다.

 또 현재의 구조가 부득이한 이유로 반드시 필요하다면, 제 생각에 뷰와 컨트롤러를 재사용하기 위해서는, 이러한 관계들이 지금처럼 DataManager라고 하는 구체타입이 아니라, 프로토콜을 통해서 언제든지 교체가 가능하면 좋을 것 같아요!

 읽어주셔서 감사합니다😅 여러분은 어떻게 생각하시는지 궁금하네요!


<br>

---

<br>

> 앱을 실행했을 때, selectFirstMemo()가 실행되면서 첫번째 셀이 선택되고 해당 셀의 내용을 detailView에 보여줘야합니다.
> 원래는 MemoListViewController의 viewDidLoad() 에서 selectFirstMemo()를 호출주었습니다.
> 그런데 List의 셀은 선택이 되지만, 선택된 셀의 상세페이지는 보이지 않는 문제가 발생했습니다.
> 반대로 MemoDetailViewController에서 viewDidLoad() 에서 selectFirstMemo()를 호출하면
> 선택된 셀의 상세페이지는 보이지만 List의 셀은 선택이 되지 않는 문제가 발생합니다.
> 해결은 했지만 왜 정확히 viewDidAppear에서는 되는건지는 알아내지 못했습니다. ㅠㅠ
> dataManager.selectFirstMemo()를 viewDidLoad에서 호출하면 왜 저희가 예상한대로 작동을 하지 않는지 궁금합니다.

 음 먼저 `DetailVC에서 호출하면 detailDelegate만 등록되고 listDelegate는 nil` 이 부분은 검증이 된 것인지 잘 모르겠네요!
 현재의 코드 구조에서는 MemoListViewController가 아니라 SplitViewController에서 문제점을 찾으셔야해요! 

 1. SplitViewController initializer가 호출
 2. SplitViewController 메모리에 적재
 3. SplitViewController viewDidLoad가 호출
 4. SplitViewController viewDidLoad 내부에서 lazy var인 listViewController와 detailViewController에 접근함으로써 각각의 initializer가 호출
 5. 코드의 순서상 listViewController initializer가 먼저 호출
 6. listViewController 메모리에 적재
 7. listViewController viewDidLoad가 호출
 8. 코드의 순서상 detailViewController initializer가 호출
 9. detailViewController 메모리에 적재
 10. detailViewController viewDidLoad가 호출

 저는 이렇게 읽히네요

> 처음에는 view context 에서 메모를 삭제하고나서 viewContext.delete(memoToDelete)
> 바로 view context를 save 한 후 saveViewContext()
> MemoDataManager가 가지고 있는 memos 배열에서도 삭제하였습니다.
> 그랬더니 Crash가 나며 앱이 중지되었습니다.
> MemoDataManager가 가지고 있는 memos 배열에서도 삭제해준 후 view context를 save 하는 방법으로
> 이렇게 순서를 바꿔서 해결하긴 했는데 Crash가 나는 이유가 궁금합니다.
> 저희가 생각하기로는 viewContext에서는 삭제가 되었지만 MemoDataManager가 가지고 있는 memos 배열에는 데이터가 남아있어서 viewContext를 저장할때 정보가 일치하지 않아서 생기는 문제라고 이해했는데 저희 생각이 적절한지 궁금합니다!

말씀하신 대로 코드를 돌려보니 크래시가 나는 걸 확인했는데요. 에러를 읽어보니 UITableView 관련 에러더라구요
<img width="881" alt="스크린샷 2022-02-23 오후 3 54 16" src="https://user-images.githubusercontent.com/83933153/155273560-7eae2e7b-0010-46da-8bf2-d2b61a621f3e.png">
delete를 하는 부분과 위 부분에서 논리적인 에러가 발생해서 그런 게 아닌가 싶네요. 앞으로는 LLDB를 찍어보시면서 어떤 함수가 호출될 때 크래시가 나는지 확인을 해보시면 잘 해결하실 수 있을 거 같습니다.

<br>

---

추가적으로 의문이 남는 부분은 코멘트를 남겼으니 확인 부탁드릴게요~

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -43,15 +44,15 @@ final class MemoListViewController: UIViewController {
+        dataManager.memos.insert(newMemo, at: 0)
         tableView.reloadData()
         tableView.selectRow(at: IndexPath(row: 0, section: 0), animated: true, scrollPosition: .top)
         delegate?.memoDetailViewController(showTextViewWith: newMemo)
```
`func setupRowSelection`에서 `dataManager.memos.isEmpty`를 하는 것이
`func addMemo`에서 `dataManager.newMemo`를 하는 것처럼 처리가 되면
dataManager.memos에 대한 처리가 더욱 안전해지고, 해당 컨트롤러에서는 dataManager에 대해서 더 몰라도 될 것 같아요.
현재의 memos의 접근제어자가 internal인데, 지금처럼 memos에 대해서 실제로 insert를 하는 부분이 컨트롤러에 있는 건 비즈니스 로직이 컨트롤러에 있다는 생각이 들고, DataManager가 managing을 제대로 하고 있지 않는 것 같아요🤔 DataManager가 실제로 data만 managing을 할 수 있도록 개선되면 좋겠습니다!

na-young-kwon replied:
수박이 우려하시는 부분이 memos의 접근제어 레벨이 internal 이라는 점 이라고 생각이 되는데요

일단 해당 커밋은 예전의 커밋이라서 따로 개선할 순 없을 것 같고,
현재 남아있는 코드들은 추후에 리팩토링하면서 dataManager가 memos를 private하게 가지고 data만 매니징 하는 쪽으로 개선해보겠습니다

siwonkim0 committed:
refactor: MemoDataManager에 있던 delegate 관련 로직 SplitVC로 이동 #8

Allie-e committed:
refactor: ListVC, DetailVC의 delegate를 SplitVC으로 지정 #8

- ListVC, DetailVC delegate protocol 및 메서드 구현

na-young-kwon committed:
refactor: ListVC, DetailVC에 있는 dataManager 삭제 #8

- MemoDataManager의 memos프로퍼티 private(set)으로 변경
- MemoListViewControllerDelegate 프로포콜에numberOfMemos, getMemo() 메서드 추가
- 필요없는 이니셜라이저 삭제
- 메모를 삭제하는 메서드 로직 수정

siwonkim0 committed:
refactor: MemoDataManagable 프로토콜 생성하여 SplitViewController에 생성자주입 #8

- memos 배열 private 로 변경
- splitViewController 이니셜라이저 생성하여 파라미터로 style과 dataManager 받도록 구현

Allie-e commented:
### 최종 설계

1. DataManager는 UI와 관련된 역할을 하지 않는다
2. SplitVC만 DataManager를 갖는다
3. ListVC / DetailVC에서는 데이터와 관련된 작업을 하지 않는다

리뷰를 읽고 ListVC / DetailVC가 DataManager를 꼭 알아야 하는지에 대해서 많은 고민을 했습니다. 확실이 DataManager를 아는 타입이 많아지면 그 만큼 관리가 어려워질 것 같아요. 그래서 수박이 피드백 해주신대로 SplitVC만 DataManager를 알고있도록 수정해주었어요. (그마저도 Protocol을 만들어서 관계를 느슨하게 해주었어요.)

기존에 ListVC / DetailVC에서 Model 관련 로직들을 모두 없앴습니다. 두 뷰컨은 딱 View를 Controller하는 역할만 하고 있어요. 

또, MemoDataManager에 남아있던 UI관련 로직들도 모두 삭제했습니다.

그 중간에서 SplitVC가 DataManager와 자식 뷰컨들을 중개해주고 있습니다.

<img width="891" alt="image" src="https://user-images.githubusercontent.com/83864058/155499820-68185f05-65c2-4c2a-9097-273c3a0d82ce.png">

### ListVC, DetailVC의 delegate = SplitVC

listViewController와 detailViewController의 delegate를 SplitVC로 지정하여 DataManager의 데이터에 따라 SplitVC가 ListVC와 DetailVC가 할 일을 대신 해주도록 구현하였습니다. 

### SOLID - DIP 원칙 (의존관계 역전 원칙)

상위레벨 모듈은 하위레벨 모듈에 의존하면 안된다는 DIP 원칙에 따라, MemoDataManagable 프로토콜을 정의하여 SplitVC이 생성될 때 MemoDataManagable 프로토콜을 준수하는 어떤 타입이라도 주입될 수 있도록 해주었습니다!

```swift
protocol MemoDataManagable {
    ...
}

extension MemoDataManager: MemoDataManagable {
    ...
}
```

```swift
final class SplitViewController: UISplitViewController {
    private let dataManager: MemoDataManagable
	...
    init(style: UISplitViewController.Style, dataManager: MemoDataManagable) {
        self.dataManager = dataManager
        super.init(style: style)
		}
	...
}
```

덕분에 MVC패턴에 대해 감을 익힐 수 있는 좋은 기회였습니다

이전 코드보다 훨씬 더 역할 분리가 잘 되어있는 것 같아요

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoDetailViewController.swift:
```diff
@@ -1,12 +1,11 @@
 import UIKit
 
 protocol MemoDetailViewControllerDelegate: AnyObject {
-    func deleteMemo(at indexPath: IndexPath?)
+    func deleteSelectedMemo()
     func updateEditedMemo(title: String, body: String, lastModified: Date)
 }
 
```
MemoDetailViewControllerDelegate가 indexPath가 필요없듯이, MemoListViewControllerDelegate도 indexPath가 필요없지 않을까요? 🤔

siwonkim0 replied:
MemoDetailViewControllerDelegate에서는 메모 상세페이지에서 더보기 버튼을 선택하여 메모를 삭제하는 상황입니다. 따라서 메모 리스트에서 선택되어있는 셀을 가져오는 메서드로 indexPath를 알 수 있었습니다. 하지만 MemoListViewControllerDelegate에서는 셀을 스와이프해서 삭제하는 상황이라서 스와이프를 하면 셀 선택이 풀리기때문에 선택된 셀을 가져올 수 없어서 부득이하게 trailingSwipeActionsConfigurationForRowAt 메서드에서 현재의 indexPath를 전달해주도록 했습니다.

soo941226 commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -107,16 +99,19 @@ extension MemoListViewController: UITableViewDelegate {
     }
     
```
67번 줄에 느낌표가 있네요~~ 느낌표는 없도록 해주세요!

나아가서 memo를 언랩핑할 필요성이 있을까 라는 생각도 드네요.

Allie-e replied:
당장 생각나는 방법을 시도해보았는데, 아래방법은 너무 물음표살인마 같아서,,
<img width="955" alt="image" src="https://user-images.githubusercontent.com/83864058/155531258-d4dae23e-6c8f-418e-b7be-dc63ae81bffd.png">

이 방법으로 수정해보았습니다.. 더 좋은 방법 있을까요?ㅎ
<img width="826" alt="image" src="https://user-images.githubusercontent.com/83864058/155531592-d6c3013b-7013-43b2-a113-faecb9e379de.png">

soo941226 replied:
전 첫번째를 생각했는데 말씀을 듣고 생각해보니 오히려 가독성이 떨어지네요. 현재의 코드가 훨씬 좋은 거 같아요👍

soo941226 commented on CloudNotes/CloudNotes/Controller/SplitViewController.swift:
```diff
@@ -2,8 +2,8 @@ import UIKit
-    private lazy var detailViewController = MemoDetailViewController(dataManager: dataManager)
+    private let listViewController = MemoListViewController()
+    private let detailViewController = MemoDetailViewController()
     
```
굳이 lazy할 필요는 없죠! 👍👍

soo941226 commented on CloudNotes/CloudNotes/Model/MemoDataManager.swift:
```diff
@@ -1,10 +1,28 @@
     private let persistentContainer: NSPersistentContainer
-    private lazy var viewContext: NSManagedObjectContext = persistentContainer.viewContext
+    private lazy var viewContext = persistentContainer.viewContext
+    private (set) var memos = [Memo]()
```
공식문서에서는 private(set)이라고 쓰긴 하네요👀 세분이 합의를 하신 부분이라면 괜찮습니다!

Allie-e replied:
수정과정을 거치면서 (set) 키워드 자체를 삭제해주었습니다! 다음에 사용할 일이 있으면 공식문서에 따라 붙여서 작성해줘야겠네요! 😅

soo941226 commented on CloudNotes/CloudNotes/Model/MemoDataManager.swift:
```diff
@@ -1,26 +1,26 @@
 import CoreData
 
+protocol MemoDataManagable {
+    var isEmpty: Bool { get }
+    var numberOfMemos: Int { get }
+    func getCurrentMemo(for indexPath: IndexPath) -> Memo
+    func insertMemo(at index: Int)
+    func updateMemo(id: UUID, title: String, body: String, lastModified: Date)
+    func deleteMemo(id: UUID)
+}
```
👏👏👏👏👏👏👏👏👏👏👏👏👏👏👏
저도 생각을 못하고 있던 요소인데 너무 좋습니다!!!!! 진짜 짱이네요😮👍👍

soo941226 commented:
정말 많이 고생하셨습니다 여러분!! 

제가 문제점이라고 여기던 각각의 역할에 대한 부분은 전부 개선되었습니다. 나머지는 중복이 되는 내용을 어떻게 줄일지에 대한 고민인 것 같아요.

머지를 하기 전에 마지막으로 질문을 드리고 싶은 게 하나 있습니다! 지금은 어떤 일이 누구의 책임인지를 따지고 리팩토링을 했는데요. 이렇게 각각의 타입이 점점 더 분명한 역할을 갖게 되었을 때 얻는 장점이 뭐라고 생각하시나요? 이전의 코드와 기능적으로는 달라지지 않았고, 이전의 코드도 기능적으로는 완벽했음에도 불구하고 말이죠...🤔

Allie-e committed:
refactor: 강제언래핑, 코드 위치 수정 및 필요없는 주석 삭제 #8

na-young-kwon commented:
#### 이렇게 각각의 타입이 점점 더 분명한 역할을 갖게 되었을 때 얻는 장점이 뭐라고 생각하시나요? 

각 타입의 역할이 명확하게 분리되면
1. 어떤 부분을 고쳐야 하는지 금방 파악할 수 있고
2. 따라서 유지보수가 용의해지고
3. 의존성이 없어지기 때문에 재사용성이 높아지고
4. 특정 타입에 대한 수정이 다른 타입에 영향을 주지 않는다는 점
이 장점이라고 생각합니다. : )

na-young-kwon committed:
docs: Update README
